### PR TITLE
shouldFail with message

### LINKED ISSUE #1502: Add 'shouldFailWithMessage' test helper
**🧐 Motivation**
Exactly as described in issue's title :).

**📝 Details**
First of all, thanks for prividing this awesome framework. It makes creating secure smart contracts actually possible!

I've been importing functions from `openzeppelin-solidity/tests/helpers` to use in my own tests for a couple of weeks. They work perfectly so there's no need to craft my own. Helpers in `shouldFail.js` have been particularly useful.

Aside, from the canned `shouldFail[reverting,throwing,outOfGas]` helpers, I think the generic `shouldFailWithMessage` is flexible and can be used to check for different failure reasons. However, it is not exported by the module.

I suggest exporting it so it could be used externally when neither of `reverting, throwing, outOfGas` fill the bill.

I will follow up with a one liner PR to introduce that change.

Thanks!

### PR DESCRIPTION
**Pull request by remon-nashid**
Fixes #1502

- Exports `shouldFailWithMessage` as `shouldFail.reverting.withMessage`.
- Adds necessary unit tests.

### CODE REVIEW

File: test/helpers/test/shouldFail.test.js [hunk 1]
```diff
@@ -92,4 +96,26 @@ describe('shouldFail', function () {
...
+    });
+
+    it('accepts require() revert with an expected reason', async function () {
+      await shouldFail.reverting.withMessage(this.failer.failRequirementVocally(), 'Unsatisfied');
+    });
+
+    it('rejects require() revert with an un expected reason', async function () {
+      await assertFailure(shouldFail.reverting.withMessage(this.failer.failRequirementVocally(), 'Unhappy'));
+    });
+  })
```
Comment (nventuro) [inline test/helpers/test/shouldFail.test.js]:
```suggestion
  });
```

File: contracts/mocks/Failer.sol [hunk 2]
```diff
@@ -19,4 +23,12 @@ contract Failer {
             array.push(i);
         }
     }
+
+    function failRequirement() public pure {
```
Comment (nventuro) [inline contracts/mocks/Failer.sol]: Do we actually need this, considering we already have `failWithRevert()`?

File: test/helpers/test/shouldFail.test.js [hunk 3]
```diff
@@ -92,4 +96,26 @@ describe('shouldFail', function () {
       await assertFailure(shouldFail.outOfGas(this.failer.failWithThrow()));
     });
   });
+
+  describe('reverting.withMessage', function () {
```
Comment (nventuro) [inline test/helpers/test/shouldFail.test.js]: Can we nest this inside the `reverting` `describe` block?

File: test/helpers/test/shouldFail.test.js (hunk 1 above)
Comment (remon-nashid) [reply test/helpers/test/shouldFail.test.js]: Oops sorry!

File: test/helpers/test/shouldFail.test.js (hunk 3 above)
Comment (remon-nashid) [reply test/helpers/test/shouldFail.test.js]: Sure will do

File: contracts/mocks/Failer.sol (hunk 2 above)
Comment (remon-nashid) [reply contracts/mocks/Failer.sol]: I realise that and `assert` and  `require` fail similarly. However, I thought that providing an additional test for the later could help *document* that `withMessage` is useful for both types of error-handling.

**Comment by remon-nashid** [comment]
> I'm also wondering if there's a way to detect if the current ganache version supports error reasons, and warn users of the helpers if it doesn't.

Interesting! I'm looking into that.

File: contracts/mocks/Failer.sol (hunk 2 above)
Comment (nventuro) [reply contracts/mocks/Failer.sol]: Hm, I guess that makes sense. Alright then!

**Comment by nventuro** [comment]
>I have no idea how hard it would be to update the fork to 6.1.8.

Alternatively, if we figure out how to detect the Ganache version, issuing a warning and skipping the message checks should be good enough.

**Comment by remon-nashid** [comment]
> The coverage job is failing because solidity-coverage uses a fork of Ganache. We're still running its version 6.1.2.

@frangio, I see [solidity-coverage@0.5.7](https://github.com/sc-forks/solidity-coverage/blob/b88e00c14ec676a3b356822cfe375bb30ef512ff/CHANGELOG.md#057--2018-08-07) uses a later version of `ethereumjs-testrpc-sc` that does support reason strings. Perhaps we could update `solidity-coverage` then?

> Alternatively, if we figure out how to detect the Ganache version, issuing a warning and skipping the message checks should be good enough.

@nventuro, I gave it much thought but not sure how to determine the ganache version used for _currently running_ test. We could always fetch ganache-cli version from `openzeppelin-solidity/package.json` during a test, but that will hold true only if locally installed (vs global) ganache is used. Also, it will make it harder to extract test helpers into their own package.. Perhaps there's a way to tell from javascript scope variables or by inspecting current contract's provider (once I put my finger on it).

**Comment by nventuro** [comment]
>I see solidity-coverage@0.5.7 uses a later version of ethereumjs-testrpc-sc that does support reason strings.

@remon-nashid great! let's try updating to that version here then and see if the tests pass.

I'm asking in their gitter about retrieving the version, it may be doable via an RPC, but if the tests pass I wouldn't block this PR on that.

**Comment by remon-nashid** [comment]
Coverage job still fails despite updating `solidity-coverage`. Looking...

**Comment by remon-nashid** [comment]
Latest commit includes a check for current ganache version. It warns users and skips message check if ganache-core version is less than 2.2.0.

@nventuro let me know if there's anything else I can do. I can't wait to use that helper from stock openzeppelin lib instead of my ageing fork.

**Comment by frangio** [comment]
Terrific @remon-nashid! We're a bit busy getting 2.1.0 ready, and also the end of the year, but we'll review this asap once that's done in the next few days.

This will likely go in the new [openzeppelin-test-helpers](https://github.com/OpenZeppelin/openzeppelin-test-helpers) package.

**Comment by remon-nashid** [comment]
Awesome! Thank you @frangio

**Comment by nventuro** [comment]
Thanks for all your help @remon-nashid! [`openzeppelin-test-helpers`](https://github.com/OpenZeppelin/openzeppelin-test-helpers) is now in a good enough place to receive this contribution, could you please open a PR on that repo with these same contents? I'm currently in the process of removing the helpers from this repository in order to use that package, so merging here doesn't make much sense.

Thanks once again!

**Comment by remon-nashid** [comment]
Here you go https://github.com/OpenZeppelin/openzeppelin-test-helpers/pull/8!

**Pull request closed without merging**